File: examples/kotlin/apps/backend/registry.kt
package app.protocol.examples.kotlin.apps.backend

import app.protocol.examples.kotlin.cases.tasks.task_create.TaskCreateApi
import app.protocol.examples.kotlin.cases.tasks.task_list.TaskListApi
import app.protocol.examples.kotlin.cases.tasks.task_move.TaskMoveApi
import app.protocol.examples.kotlin.core.shared.AppCaseSurfaces
import app.protocol.examples.kotlin.core.shared.AppRegistry
import app.protocol.examples.kotlin.packages.data.JsonFileStore
import app.protocol.examples.kotlin.packages.data.createDataPackage
import app.protocol.examples.kotlin.packages.data.createJsonFileStore
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonObject

data class BackendConfig(
    val port: Int = 3000,
    val dataDirectory: String? = null,
)

data class BackendRegistry(
    override val _cases: Map<String, Map<String, AppCaseSurfaces>>,
    override val _providers: Map<String, Any?>?,
    override val _packages: Map<String, Any?>?,
) : AppRegistry

private fun createTaskStoreProvider(
    taskStore: JsonFileStore<List<JsonObject>>,
): Map<String, Any?> {
    val read: suspend () -> List<JsonObject> = { taskStore.read() }
    val write: suspend (List<JsonObject>) -> Unit = { value -> taskStore.write(value) }
    val reset: suspend () -> Unit = { taskStore.reset() }
    val update: suspend (suspend (List<JsonObject>) -> List<JsonObject>) -> List<JsonObject> =
        { updater -> taskStore.update(updater) }

    return mapOf(
        "read" to read,
        "write" to write,
        "reset" to reset,
        "update" to update,
    )
}

fun createRegistry(config: BackendConfig = BackendConfig()): BackendRegistry {
    val data = createDataPackage(config.dataDirectory)
    val taskStore =
        createJsonFileStore(
            filePath = data.defaultFiles.tasks,
            fallbackData = emptyList<JsonObject>(),
            serializer = ListSerializer(JsonObject.serializer()),
        )

    return BackendRegistry(
        _cases = mapOf(
            "tasks" to mapOf(
                "task_create" to AppCaseSurfaces(api = { ctx -> TaskCreateApi(ctx) }),
                "task_list" to AppCaseSurfaces(api = { ctx -> TaskListApi(ctx) }),
                "task_move" to AppCaseSurfaces(api = { ctx -> TaskMoveApi(ctx) }),
            ),
        ),
        _providers = mapOf(
            "port" to config.port,
            "taskStore" to createTaskStoreProvider(taskStore),
        ),
        _packages = mapOf(
            "data" to data,
        ),
    )
}

File: examples/go/apps/backend/registry.go
package backend

import (
	"path/filepath"

	"app-protocol/examples/go/cases/tasks/task_create"
	"app-protocol/examples/go/cases/tasks/task_list"
	"app-protocol/examples/go/cases/tasks/task_move"
	"app-protocol/examples/go/core"
	"app-protocol/examples/go/core/shared"
	"app-protocol/examples/go/packages/data"
)

type Config struct {
	Port          int
	DataDirectory string
}

type Registry struct {
	_cases     map[string]map[string]shared.AppCaseSurfaces
	_providers shared.Dict
	_packages  shared.Dict
}

func CreateRegistry(config Config) *Registry {
	dataDirectory := config.DataDirectory
	if dataDirectory == "" {
		dataDirectory = filepath.Join("packages", "data")
	}

	dataPackage := data.New(dataDirectory)
	taskStore := data.NewJSONFileStore[[]map[string]any](dataPackage.DefaultFiles.Tasks, []map[string]any{})

	return &Registry{
		_cases: map[string]map[string]shared.AppCaseSurfaces{
			"tasks": {
				"task_create": {
					API: func(ctx any) any {
						return task_create.NewAPI(ctx.(*core.ApiContext))
					},
				},
				"task_list": {
					API: func(ctx any) any {
						return task_list.NewAPI(ctx.(*core.ApiContext))
					},
				},
				"task_move": {
					API: func(ctx any) any {
						return task_move.NewAPI(ctx.(*core.ApiContext))
					},
				},
			},
		},
		_providers: shared.Dict{
			"port":      config.Port,
			"taskStore": taskStore,
		},
		_packages: shared.Dict{
			"data": dataPackage,
		},
	}
}

func (r *Registry) Cases() map[string]map[string]shared.AppCaseSurfaces {
	return r._cases
}

func (r *Registry) Providers() shared.Dict {
	return r._providers
}

func (r *Registry) Packages() shared.Dict {
	return r._packages
}
